Fix for #1002

jbracker opened the pull request:
This is a fix for #1002. 

All it does it replace the default implementation of `Observable.combineLatestList` with the custom implementation `CombineLatestListObservable`. The custom implementation takes special care to not cause any stack issues and tries to produce as little overhead as possible.

`CombineLatestListObservable` is based on the template set by `CombineLatest2Observable` and just expands on it to handle any number of incoming observables.

Should `Observable.combineLatestList` use the specialized implementations for 1 to 6 incoming `Observable`s if possible (i.e. `CombineLatest2Observable`, `CombineLatest3Observable`, etc.)?

Avasil commented:
Thank you very much for PR @jbracker 
We are a bit busy with `3.0.0` release so we are slacking a bit with reviews but I will try to review it as soon as I can and this is something we can add in `bugfix` release.

> Should Observable.combineLatestList use the specialized implementations for 1 to 6 incoming Observables if possible (i.e. CombineLatest2Observable, CombineLatest3Observable, etc.)?

It will be hard to know unless we benchmark it so I think it's OK in the current state

alexandru commented:
A review for this ticket is on my TODO 🙂

The problem with these `Observable` operators is that they operate with concurrent side effects.

jbracker commented:
Glad to hear that this is making progress :)

> The problem with these Observable operators is that they operate with concurrent side effects.

That's why I based the implementation on `CombineLatest2Observable`.

jakoschiko commented on monix-reactive/shared/src/main/scala/monix/reactive/Observable.scala:
```diff
@@ -5859,14 +5859,10 @@ object Observable extends ObservableDeprecatedBuilders {
     * returning a new observable that generates sequences.
```
The doc mentions that the implemenation is based on combineLatestMap2. That's not true anymore.

Avasil commented on monix-reactive/shared/src/main/scala/monix/reactive/internal/builders/CombineLatestListObservable.scala:
```diff
@@ -0,0 +1,164 @@
+    // MUST BE synchronized by `lock`
+    val elems: mutable.ArraySeq[A] = new mutable.ArraySeq(numberOfObservables)
+    // MUST BE synchronized by `lock`
+    val hasElems: Array[Boolean] = new Array(numberOfObservables)
```
Could be `Array.fill(numberOfObservables)(false)`

Avasil commented on monix-reactive/shared/src/main/scala/monix/reactive/internal/builders/CombineLatestListObservable.scala:
```diff
@@ -0,0 +1,164 @@
+    // NOTE: We use arrays and other mutable structures here to be as performant as possible.
+
+    // MUST BE synchronized by `lock`
+    val observables: Array[Observable[(A, Int)]] = new Array(numberOfObservables)
```
I wonder if we could store `(Int, Observable[A])` instead to not create and unpack a tuple on each `onNext`?

Avasil commented on monix-reactive/shared/src/main/scala/monix/reactive/internal/builders/CombineLatestListObservable.scala:
```diff
@@ -0,0 +1,164 @@
+              hasElems(index) = true
+            }
+
+            if (hasElems.forall(identity)) {
```
We could probably set a flag the first time it is true so we don't have to check full array each time

jakoschiko replied:
Or we store the number of `true`s in an additional val, so we never have to check the full array.

jbracker commented:
@Avasil @jakoschiko I have addressed some of the comments:

- Comments have been updated. I did not see any harm leaving the class comment for `CombineLatestListObservable`.
- `toArray` is not an option because we don't have a `ClassTag` for `Observable[A]` and I don't think we wan't one there. I could make it an ArraySeq or a Vector though if you think that would be preferable.
- Using `Array.fill` now for initialization of `hasElems`.
- Yes, you are right I don't need to map the index into the observable. Having it on the outside is sufficient.
- I created the counter `hasElemCount` to allow fast checks.

oleg-py commented on monix-reactive/shared/src/main/scala/monix/reactive/internal/builders/CombineLatestListObservable.scala:
```diff
@@ -0,0 +1,163 @@
+    observables.foreach { case (obs, index) =>
+      composite += obs.unsafeSubscribeFn(new Subscriber[A] {
+        implicit val scheduler: Scheduler = out.scheduler
+
```
Correct me if I wrong, but why do we need these tuples in `observables`? It seems that `index` is only used here, so one should be able to drop these allocations just by doing:
```var i = 0
observables.foreach { obs =>
  val index = i // close over proper index
  i += 1
```
and thus you don't need to reallocate collection at all (can just use original `obss` coming to you)

jakoschiko replied:
Or
```
observables.iterator.zipWithIndex.map { case (obs, index) =>
    ???
}}}
```

oleg-py replied:
@jakoschiko that is indeed an option, but it's only halfway - you skip the allocation of an extra collection, but not the tuples for each instance.

oleg-py commented on monix-reactive/shared/src/main/scala/monix/reactive/internal/builders/CombineLatestListObservable.scala:
```diff
@@ -0,0 +1,164 @@
+import scala.util.Success
+import scala.util.control.NonFatal
+
+/** Only used in [[Observable.combineLatestList()]]. Is tested by `CombineLatestListSuite`. */
```
Yeah, we don't really mention tests in comments, but we do say "implementation for `Observable.combineLatestList`"

oleg-py commented on monix-reactive/shared/src/main/scala/monix/reactive/internal/builders/CombineLatestListObservable.scala:
```diff
@@ -0,0 +1,162 @@
+import scala.util.control.NonFatal
+
+/** Only used in [[Observable.combineLatestList()]]. Is tested by `CombineLatestListSuite`. */
+private[reactive] final class CombineLatestListObservable[A, +R](observables: Seq[Observable[A]])(f: Seq[A] => R)
```
I think we can get away without the `f` here.

`combineLatestMapX` helps avoid tuple allocations and induced primitive boxing, but we probably aren't buying much over `combineLatestList(...).map` as you do in your tests.

If you go with dropping it, you won't need to guard against user errors in f, so:
```scala
def rawOnNext(as: Seq[A]): Future[Ack] = if (isDone) Stop else out.onNext(as)
```

Avasil commented:
Many thanks @jbracker :)
